src/rss_etl.py

```diff
@@ -0,0 +1,175 @@
+from supabase import create_client, Client
+import feedparser
+import os
+import re
+from enum import Enum
+from dataclasses import dataclass
+
+"""
+File for extracting data from the RSS feeds of the PA Senate and House of Representatives
+and loading it into Supabase.
+
+The database tables updated by this script are:
+    Bills
+    Revisions
+"""
+
+# from dotenv import load_dotenv
+# load_dotenv()  # for local secret management with .env file
+
+senate_rss_feed = "https://www.legis.state.pa.us/WU01/LI/RSS/SenateBills.xml"
+house_rss_feed = "https://www.legis.state.pa.us/WU01/LI/RSS/HouseBills.xml"
+
+# sb_api_url = os.getenv("SUPABASE_API_URL")  # local secret management with .env file
+# sb_api_key = os.getenv("SUPABASE_API_KEY")  # local secret management with .env file
+sb_api_url = os.environ["SUPABASE_API_URL"]  # github actions secret management
+sb_api_key = os.environ["SUPABASE_API_KEY"]  # github actions secret management
+
+
+class LegislativeBody(Enum):
+    SENATE = "Senate"
+    HOUSE = "House"
+
+
+REVISIONS_OUTPUT_DICT = {
+    "bill_internal_id": "",
+    "printer_no": "",
+    "full_text_link": "",
+    "publication_date": "",
+    "revision_guid": "",
+    "description": "",
+}
+
+BILL_OUTPUT_DICT = {
+    "bill_number": "",
+    "legislative_session": "",
+    "session_type": "",
+    "chamber": "",
+    "legislative_id": "",
+}
+
+
+class BillIdentifier:
+    """
+    Extracts unique information from a bill's guid and embeds them in discrete properties
+    """
+
+    @staticmethod
+    def extract_bill_guid(revision_guid):
+        return re.search("BillGuid=(.*)", revision_guid).group(1)
+
+    @staticmethod
+    def determine_chamber(chamber_letter):
+        if chamber_letter == "H":
+            return "House"
+        elif chamber_letter == "S":
+            return "Senate"
+        else:
+            raise Exception("Invalid chamber letter in guid")
+
+    def __init__(self, guid):
+        self.guid = guid
+        self.bill_guid = re.search(r"(\d{4}\d[H|S]B\d+)P\d+", guid).group(1)
```

Comment:
This regex threw an exception for me while trying this out. The above re.search call returns "None" if the regex fails and the rss feed gave what I am guessing is a revision instead of a bill so it had an 'R' where a 'B' was expected. Is the BillIdentifier class meant to handle both bills and revisions?

Reply:
For my local testing I changed the regexes to something like `(\d{4}\d[H|S](?:B|R)\d+)P\d+` just to not mess with the groups. Also the `[H|S]` matches an 'H', an 'S', or a '|'. With regexes everything in square brackets is treated as a literal character (expect '^' and '-' in certain cases)

Reply:
> This regex threw an exception for me while trying this out. The above re.search call returns "None" if the regex fails and the rss feed gave what I am guessing is a revision instead of a bill so it had an 'R' where a 'B' was expected. Is the BillIdentifier class meant to handle both bills and revisions?

All RSS entries should be bill revisions, rather than bills itself. This is in part because bills have a one-to-many relationship with revisions -- one bill can have multiple revisions, but the opposite is not true. A bill must have at least one revision.

The R in that case did not refer to a Revision, but a _Resolution_. This is unclear in my code, and I am updating it accordingly as part of my attempts to explain the regex better overall.

> For my local testing I changed the regexes to something like `(\d{4}\d[H|S](?:B|R)\d+)P\d+` just to not mess with the groups. Also the `[H|S]` matches an 'H', an 'S', or a '|'. With regexes everything in square brackets is treated as a literal character (expect '^' and '-' in certain cases)

Excellent catch! I've corrected this in my own code as well.